File: prototypes/graph-pilot/fx-util.js
/* ============================================================================
 * Shared FX-rate resolution for the DAG's what-if tool.
 * Default matches engine/constants.js's CONST.FX.INR_PER_USD (83.0 INR/USD) —
 * every node file used to hard-copy this literal into its own inrToUsd/
 * moneyFromInr helpers (see docs/DAG_MIGRATION_TRACKER.md's FX note). Those
 * helpers now read the rate through fxRate(ctx) instead, so dragging the
 * what-if FX slider (ctx.fxRateOverride, set by monitor-next/lib/dag-adapter.js)
 * re-prices every INR<->USD conversion in the graph from one place. With no
 * override present, behavior is byte-identical to the old bare literal.
 * ==========================================================================*/
var DEFAULT_FX_RATE = 83.0;

function fxRate(ctx) {
  var v = ctx && ctx.fxRateOverride;
  var n = Number(v);
  return (v !== undefined && v !== null && !isNaN(n) && n > 0) ? n : DEFAULT_FX_RATE;
}

module.exports = { fxRate: fxRate, DEFAULT_FX_RATE: DEFAULT_FX_RATE };

File: dag_py/src/wising_dag/core/fx_util.py
"""Shared FX-rate resolution — port of prototypes/graph-pilot/fx-util.js.

Default matches archive/engine-frozen/constants.js's CONST.FX.INR_PER_USD
(83.0 INR/USD). ctx.fx_rate_override (set by the what-if tool) re-prices
every INR<->USD conversion in the graph from this one place; with no
override present, behavior is byte-identical to the bare literal.
"""
from __future__ import annotations

DEFAULT_FX_RATE = 83.0


def fx_rate(ctx: dict) -> float:
    v = ctx.get("fxRateOverride") if ctx else None
    if v is None:
        return DEFAULT_FX_RATE
    try:
        n = float(v)
    except (TypeError, ValueError):
        return DEFAULT_FX_RATE
    return n if n > 0 else DEFAULT_FX_RATE
